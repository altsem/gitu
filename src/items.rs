use crate::config::Config;
use crate::git::diff::Delta;
use crate::git::diff::Diff;
use crate::git::diff::Hunk;
use crate::Res;
use git2::Commit;
use git2::Oid;
use git2::Repository;
use ratatui::style::Style;
use ratatui::text::Line;
use ratatui::text::Span;
use regex::Regex;
use std::borrow::Cow;
use std::iter;
use std::path::PathBuf;
use std::rc::Rc;

#[derive(Default, Clone, Debug)]
pub(crate) struct Item {
    pub(crate) id: Cow<'static, str>,
    pub(crate) display: Line<'static>,
    pub(crate) section: bool,
    pub(crate) default_collapsed: bool,
    pub(crate) depth: usize,
    pub(crate) unselectable: bool,
    pub(crate) target_data: Option<TargetData>,
}

#[derive(Clone, Debug)]
pub(crate) enum TargetData {
    AllStaged,
    AllUnstaged,
    AllUntracked(Vec<PathBuf>),
    Branch(String),
    Commit(String),
    Delta(Delta),
    File(PathBuf),
    Hunk(Rc<Hunk>),
    HunkLine(Rc<Hunk>, usize),
    Stash { commit: String, id: usize },
}

pub(crate) fn create_diff_items<'a>(
    config: Rc<Config>,
    diff: &'a Diff,
    depth: &'a usize,
    default_collapsed: bool,
) -> impl Iterator<Item = Item> + 'a {
    diff.deltas.iter().flat_map(move |delta| {
        let target_data = TargetData::Delta(delta.clone());
        let config = Rc::clone(&config);

        iter::once(Item {
            id: delta.file_header.to_string().into(),
            display: Line::styled(
                format!(
                    "{}   {}",
                    format!("{:?}", delta.status).to_lowercase(),
                    delta.new_file.to_string_lossy()
                ),
                &config.style.file_header,
            ),
            section: true,
            default_collapsed,
            depth: *depth,
            target_data: Some(target_data),
            ..Default::default()
        })
        .chain(
            delta
                .hunks
                .iter()
                .cloned()
                .flat_map(move |hunk| create_hunk_items(Rc::clone(&config), hunk, *depth + 1)),
        )
    })
}

fn create_hunk_items(
    config: Rc<Config>,
    hunk: Rc<Hunk>,
    depth: usize,
) -> impl Iterator<Item = Item> {
    let target_data = TargetData::Hunk(Rc::clone(&hunk));

    iter::once(Item {
        id: hunk.format_patch().into(),
        display: Line::styled(hunk.header.clone(), &config.style.hunk_header),
        section: true,
        depth,
        target_data: Some(target_data),
        ..Default::default()
    })
    .chain(format_diff_hunk_items(depth + 1, hunk))
}

fn format_diff_hunk_items(depth: usize, hunk: Rc<Hunk>) -> Vec<Item> {
    hunk.content
        .lines
        .iter()
        .enumerate()
        .map(|(i, line)| Item {
            display: replace_tabs_with_spaces(line.clone()),
            unselectable: line
                .spans
                .first()
                .is_some_and(|s| s.content.starts_with(' ')),
            depth,
            target_data: Some(TargetData::HunkLine(Rc::clone(&hunk), i)),
            ..Default::default()
        })
        .collect()
}

fn replace_tabs_with_spaces(line: Line<'_>) -> Line<'_> {
    let spans = line
        .spans
        .iter()
        .cloned()
        .map(|span| Span::styled(span.content.replace('\t', "    "), span.style))
        .collect::<Vec<_>>();

    Line { spans, ..line }
}

pub(crate) fn stash_list(config: &Config, repo: &Repository, limit: usize) -> Res<Vec<Item>> {
    let style = &config.style;

    Ok(repo
        .reflog("refs/stash")?
        .iter()
        .enumerate()
        .map(|(i, stash)| -> Res<Item> {
            let spans = itertools::intersperse(
                iter::once(Span::styled(format!("stash@{i}"), &style.hash)).chain([stash
                    .message()
                    .unwrap_or("")
                    .to_string()
                    .into()]),
                Span::raw(" "),
            )
            .collect::<Vec<_>>();

            Ok(Item {
                id: stash.id_new().to_string().into(),
                display: Line::from(spans),
                depth: 1,
                target_data: Some(TargetData::Stash {
                    commit: stash.id_new().to_string(),
                    id: i,
                }),
                ..Default::default()
            })
        })
        .map(|result| match result {
            Ok(item) => item,
            Err(err) => Item {
                id: err.to_string().into(),
                display: err.to_string().into(),
                ..Default::default()
            },
        })
        .take(limit)
        .collect::<Vec<_>>())
}

pub(crate) fn log(
    config: &Config,
    repo: &Repository,
    limit: usize,
    rev: Option<Oid>,
    msg_regex: Option<Regex>,
) -> Res<Vec<Item>> {
    let style = &config.style;
    let mut revwalk = repo.revwalk()?;
    if let Some(r) = rev {
        revwalk.push(r)?;
    } else if revwalk.push_head().is_err() {
        return Ok(vec![]);
    }

    let references = repo
        .references()?
        .filter_map(Result::ok)
        .filter_map(
            |reference| match (reference.peel_to_commit(), reference.shorthand()) {
                (Ok(target), Some(name)) => {
                    if name.ends_with("/HEAD") || name.starts_with("prefetch/remotes/") {
                        return None;
                    }

                    let style: Style = if reference.is_remote() {
                        &style.remote
                    } else if reference.is_tag() {
                        &style.tag
                    } else {
                        &style.branch
                    }
                    .into();

                    Some((target, Span::styled(name.to_string(), style)))
                }
                _ => None,
            },
        )
        .collect::<Vec<(Commit, Span)>>();

    let items: Vec<Item> = revwalk
        .map(|oid_result| -> Res<Option<Item>> {
            let oid = oid_result?;
            let commit = repo.find_commit(oid)?;
            let short_id = commit.as_object().short_id()?.as_str().unwrap().to_string();

            let spans = itertools::intersperse(
                iter::once(Span::styled(short_id, &style.hash))
                    .chain(
                        references
                            .iter()
                            .filter(|(commit, _)| commit.id() == oid)
                            .map(|(_, name)| name.clone()),
                    )
                    .chain([commit.summary().unwrap_or("").to_string().into()]),
                Span::raw(" "),
            )
            .collect::<Vec<_>>();

            if let Some(re) = &msg_regex {
                if !re.is_match(commit.message().unwrap_or("")) {
                    return Ok(None);
                }
            }

            Ok(Some(Item {
                id: oid.to_string().into(),
                display: Line::from(spans),
                depth: 1,
                target_data: Some(TargetData::Commit(oid.to_string())),
                ..Default::default()
            }))
        })
        .filter_map(|result| match result {
            Ok(item) => item,
            Err(err) => Some(Item {
                id: err.to_string().into(),
                display: err.to_string().into(),
                ..Default::default()
            }),
        })
        .take(limit)
        .collect();

    if items.is_empty() {
        Ok(vec![Item {
            display: Line::raw("No commits found"),
            ..Default::default()
        }])
    } else {
        Ok(items)
    }
}

pub(crate) fn blank_line() -> Item {
    Item {
        display: Line::raw(""),
        depth: 0,
        unselectable: true,
        ..Default::default()
    }
}
