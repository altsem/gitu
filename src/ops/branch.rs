use super::{selected_rev, Action, OpTrait};
use crate::{
    app::{App, PromptParams, State},
    error::Error,
    git::{get_current_branch_name, is_branch_merged},
    item_data::{ItemData, RefKind},
    menu::arg::Arg,
    term::Term,
    Res,
};
use std::{process::Command, rc::Rc};

pub(crate) fn init_args() -> Vec<Arg> {
    vec![]
}

pub(crate) struct Checkout;
impl OpTrait for Checkout {
    fn get_action(&self, _target: &ItemData) -> Option<Action> {
        Some(Rc::new(move |app: &mut App, term: &mut Term| {
            let rev = app.prompt(
                term,
                &PromptParams {
                    prompt: "Checkout",
                    create_default_value: Box::new(selected_rev),
                    ..Default::default()
                },
            )?;

            checkout(app, term, &rev)?;
            Ok(())
        }))
    }

    fn display(&self, _state: &State) -> String {
        "Checkout branch/revision".into()
    }
}

fn checkout(app: &mut App, term: &mut Term, rev: &str) -> Res<()> {
    let mut cmd = Command::new("git");
    cmd.args(["checkout", rev]);

    app.close_menu();
    app.run_cmd(term, &[], cmd)?;
    Ok(())
}

pub(crate) struct CheckoutNewBranch;
impl OpTrait for CheckoutNewBranch {
    fn get_action(&self, _target: &ItemData) -> Option<Action> {
        Some(Rc::new(|app: &mut App, term: &mut Term| {
            let branch_name = app.prompt(
                term,
                &PromptParams {
                    prompt: "Create and checkout branch:",
                    ..Default::default()
                },
            )?;

            checkout_new_branch_prompt_update(app, term, &branch_name)?;
            Ok(())
        }))
    }

    fn display(&self, _state: &State) -> String {
        "Checkout new branch".into()
    }
}

fn checkout_new_branch_prompt_update(app: &mut App, term: &mut Term, branch_name: &str) -> Res<()> {
    let mut cmd = Command::new("git");
    cmd.args(["checkout", "-b", branch_name]);

    app.close_menu();
    app.run_cmd(term, &[], cmd)?;
    Ok(())
}

pub(crate) struct Delete;
impl OpTrait for Delete {
    fn get_action(&self, target: &ItemData) -> Option<Action> {
        let default = match target {
            ItemData::Reference {
                kind: RefKind::Branch(branch),
                ..
            } => Some(branch.clone()),
            _ => None,
        };

        Some(Rc::new(move |app: &mut App, term: &mut Term| {
            let default = default.clone();

            let branch_name = app.prompt(
                term,
                &PromptParams {
                    prompt: "Delete",
                    create_default_value: Box::new(move |_| default.clone()),
                    ..Default::default()
                },
            )?;

            delete(app, term, &branch_name)?;
            Ok(())
        }))
    }

    fn display(&self, _state: &State) -> String {
        "Delete branch".into()
    }
}

pub fn delete(app: &mut App, term: &mut Term, branch_name: &str) -> Res<()> {
    if branch_name.is_empty() {
        return Err(Error::BranchNameRequired);
    }

    if get_current_branch_name(&app.state.repo).unwrap() == branch_name {
        return Err(Error::CannotDeleteCurrentBranch);
    }

    let mut cmd = Command::new("git");
    cmd.args(["branch", "-d"]);

    if !is_branch_merged(&app.state.repo, branch_name).unwrap_or(false) {
        app.confirm(term, "Branch is not fully merged. Really delete? (y or n)")?;
        cmd.arg("-f");
    }

    cmd.arg(branch_name);

    app.close_menu();
    app.run_cmd(term, &[], cmd)?;
    Ok(())
}
