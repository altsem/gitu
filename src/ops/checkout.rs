use super::{Action, OpTrait};
use crate::{items::TargetData, prompt::PromptData, state::State, term::Term, Res};
use derive_more::Display;
use std::{process::Command, rc::Rc};
use tui_prompts::State as _;

#[derive(Default, Clone, Copy, PartialEq, Eq, Debug, Display)]
#[display(fmt = "Checkout branch/revision")]
pub(crate) struct Checkout;
impl OpTrait for Checkout {
    fn get_action(&self, _target: Option<&TargetData>) -> Option<Action> {
        Some(Rc::new(|state: &mut State, _term: &mut Term| {
            let prompt_text = if let Some(branch_or_revision) = default_branch_or_revision(state) {
                format!("Checkout (default {}):", branch_or_revision).into()
            } else {
                "Checkout:".into()
            };

            state.prompt.set(PromptData {
                prompt_text,
                update_fn: Rc::new(checkout_prompt_update),
            });
            Ok(())
        }))
    }
}

fn checkout_prompt_update(state: &mut State, term: &mut Term) -> Res<()> {
    if state.prompt.state.status().is_done() {
        let input = state.prompt.state.value().to_string();
        let branch_or_revision = match (input.as_str(), default_branch_or_revision(state)) {
            ("", None) => "",
            ("", Some(default)) => default,
            (value, _) => value,
        };

        let mut cmd = Command::new("git");
        cmd.args(["checkout", &branch_or_revision]);

        state.run_external_cmd(term, &[], cmd)?;
        state.prompt.reset(term)?;
    }
    Ok(())
}

fn default_branch_or_revision(state: &State) -> Option<&str> {
    match &state.screen().get_selected_item().target_data {
        Some(TargetData::Branch(branch)) => Some(branch),
        Some(TargetData::Commit(commit)) => Some(commit),
        _ => None,
    }
}

#[derive(Default, Clone, Copy, PartialEq, Eq, Debug, Display)]
#[display(fmt = "Checkout new branch")]
pub(crate) struct CheckoutNewBranch;
impl OpTrait for CheckoutNewBranch {
    fn get_action(&self, _target: Option<&TargetData>) -> Option<Action> {
        Some(Rc::new(|state: &mut State, _term: &mut Term| {
            state.prompt.set(PromptData {
                prompt_text: "Create and checkout branch:".into(),
                update_fn: Rc::new(checkout_new_branch_prompt_update),
            });
            Ok(())
        }))
    }
}

fn checkout_new_branch_prompt_update(state: &mut State, term: &mut Term) -> Res<()> {
    if state.prompt.state.status().is_done() {
        let name = state.prompt.state.value().to_string();
        let mut cmd = Command::new("git");
        cmd.args(["checkout", "-b", &name]);

        state.run_external_cmd(term, &[], cmd)?;
        state.prompt.reset(term)?;
    }
    Ok(())
}
