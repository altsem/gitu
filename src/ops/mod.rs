use serde::{Deserialize, Serialize};
use tui_prompts::State as _;

use crate::{
    cmd_log::CmdLogEntry, items::TargetData, menu::Menu, prompt::PromptData, state::State,
    term::Term, Res,
};
use std::{fmt::Display, rc::Rc};

pub(crate) mod checkout;
pub(crate) mod commit;
pub(crate) mod copy_hash;
pub(crate) mod discard;
pub(crate) mod editor;
pub(crate) mod fetch;
pub(crate) mod log;
pub(crate) mod pull;
pub(crate) mod push;
pub(crate) mod rebase;
pub(crate) mod reset;
pub(crate) mod revert;
pub(crate) mod show;
pub(crate) mod show_refs;
pub(crate) mod stage;
pub(crate) mod stash;
pub(crate) mod unstage;

pub(crate) type Action = Rc<dyn FnMut(&mut State, &mut Term) -> Res<()>>;

pub(crate) trait OpTrait: Display {
    /// Get the implementation (which may or may not exist) of the Op given some TargetData.
    /// This indirection allows Gitu to show a contextual menu of applicable actions.
    fn get_action(&self, target: Option<&TargetData>) -> Option<Action>;

    /// This indicates whether the Op is meant to read and
    /// act on TargetData. Those are listed differently in the help menu.
    fn is_target_op(&self) -> bool {
        false
    }
}

#[derive(Clone, PartialOrd, Ord, PartialEq, Eq, Debug, Serialize, Deserialize)]
#[serde(rename_all = "snake_case")]
pub(crate) enum Op {
    Checkout,
    CheckoutNewBranch,
    Commit,
    CommitAmend,
    FetchAll,
    FetchElsewhere,
    LogCurrent,
    Pull,
    PullElsewhere,
    Push,
    PushElsewhere,
    RebaseAbort,
    RebaseContinue,
    RebaseElsewhere,
    ShowRefs,
    Stash,
    StashApply,
    StashIndex,
    StashWorktree,
    StashKeepIndex,
    StashPop,
    StashDrop,
    CommitFixup,
    LogOther,
    RebaseAutosquash,
    RebaseInteractive,
    ResetSoft,
    ResetMixed,
    ResetHard,
    RevertAbort,
    RevertContinue,
    RevertCommit,

    Stage,
    Unstage,
    Show,
    Discard,
    CopyHash,

    ToggleSection,
    MoveUp,
    MoveDown,
    MoveUpLine,
    MoveDownLine,
    MovePrevSection,
    MoveNextSection,
    MoveParentSection,
    HalfPageUp,
    HalfPageDown,
    ScrollDown,
    ScrollUp,

    Refresh,
    Quit,

    #[serde(untagged)]
    OpenMenu(Menu),
    #[serde(untagged)]
    ToggleArg(String),
}

impl Op {
    pub fn implementation(self) -> Box<dyn OpTrait> {
        match self {
            Op::Quit => Box::new(editor::Quit),
            Op::OpenMenu(menu) => Box::new(editor::OpenMenu(menu)),
            Op::Refresh => Box::new(editor::Refresh),
            Op::ToggleArg(name) => Box::new(editor::ToggleArg(name)),
            Op::ToggleSection => Box::new(editor::ToggleSection),
            Op::MoveDown => Box::new(editor::MoveDown),
            Op::MoveUp => Box::new(editor::MoveUp),
            Op::MoveDownLine => Box::new(editor::MoveDownLine),
            Op::MoveUpLine => Box::new(editor::MoveUpLine),
            Op::MoveNextSection => Box::new(editor::MoveNextSection),
            Op::MovePrevSection => Box::new(editor::MovePrevSection),
            Op::MoveParentSection => Box::new(editor::MoveParentSection),
            Op::HalfPageUp => Box::new(editor::HalfPageUp),
            Op::HalfPageDown => Box::new(editor::HalfPageDown),
            Op::ScrollDown => Box::new(editor::ScrollDown),
            Op::ScrollUp => Box::new(editor::ScrollUp),

            Op::Checkout => Box::new(checkout::Checkout),
            Op::CheckoutNewBranch => Box::new(checkout::CheckoutNewBranch),
            Op::Commit => Box::new(commit::Commit),
            Op::CommitAmend => Box::new(commit::CommitAmend),
            Op::FetchAll => Box::new(fetch::FetchAll),
            Op::FetchElsewhere => Box::new(fetch::FetchElsewhere),
            Op::LogCurrent => Box::new(log::LogCurrent),
            Op::Pull => Box::new(pull::Pull),
            Op::PullElsewhere => Box::new(pull::PullElsewhere),
            Op::Push => Box::new(push::Push),
            Op::PushElsewhere => Box::new(push::PushElsewhere),
            Op::RebaseAbort => Box::new(rebase::RebaseAbort),
            Op::RebaseContinue => Box::new(rebase::RebaseContinue),
            Op::RebaseElsewhere => Box::new(rebase::RebaseElsewhere),
            Op::ShowRefs => Box::new(show_refs::ShowRefs),
            Op::Stash => Box::new(stash::Stash),
            Op::StashApply => Box::new(stash::StashApply),
            Op::StashIndex => Box::new(stash::StashIndex),
            Op::StashWorktree => Box::new(stash::StashWorktree),
            Op::StashKeepIndex => Box::new(stash::StashKeepIndex),
            Op::StashPop => Box::new(stash::StashPop),
            Op::StashDrop => Box::new(stash::StashDrop),

            Op::CommitFixup => Box::new(commit::CommitFixup),
            Op::Discard => Box::new(discard::Discard),
            Op::LogOther => Box::new(log::LogOther),
            Op::RebaseAutosquash => Box::new(rebase::RebaseAutosquash),
            Op::RebaseInteractive => Box::new(rebase::RebaseInteractive),
            Op::ResetSoft => Box::new(reset::ResetSoft),
            Op::ResetMixed => Box::new(reset::ResetMixed),
            Op::ResetHard => Box::new(reset::ResetHard),
            Op::RevertAbort => Box::new(revert::RevertAbort),
            Op::RevertContinue => Box::new(revert::RevertContinue),
            Op::RevertCommit => Box::new(revert::RevertCommit),
            Op::Show => Box::new(show::Show),
            Op::Stage => Box::new(stage::Stage),
            Op::Unstage => Box::new(unstage::Unstage),
            Op::CopyHash => Box::new(copy_hash::CopyHash),
        }
    }
}

impl Display for Menu {
    fn fmt(&self, f: &mut std::fmt::Formatter<'_>) -> std::fmt::Result {
        f.write_str(match self {
            Menu::Root => "Root",
            Menu::Branch => "Branch",
            Menu::Commit => "Commit",
            Menu::Fetch => "Fetch",
            Menu::Help => "Help",
            Menu::Log => "Log",
            Menu::Pull => "Pull",
            Menu::Push => "Push",
            Menu::Rebase => "Rebase",
            Menu::Reset => "Reset",
            Menu::Revert => "Revert",
            Menu::Stash => "Stash",
        })
    }
}

pub(crate) fn create_y_n_prompt(mut action: Action, prompt: &'static str) -> Action {
    let update_fn = Rc::new(move |state: &mut State, term: &mut Term| {
        if state.prompt.state.status().is_pending() {
            match state.prompt.state.value() {
                "y" => {
                    Rc::get_mut(&mut action).unwrap()(state, term)?;
                    state.prompt.reset(term)?;
                }
                "" => (),
                _ => {
                    state
                        .current_cmd_log
                        .push(CmdLogEntry::Error("Aborted".to_string()));
                    state.prompt.reset(term)?;
                }
            }
        }
        Ok(())
    });

    Rc::new(move |state: &mut State, _term: &mut Term| {
        state.prompt.set(PromptData {
            prompt_text: format!("{} (y or n)", prompt).into(),
            update_fn: update_fn.clone(),
        });

        Ok(())
    })
}

pub(crate) fn create_prompt(
    prompt: &'static str,
    callback: fn(&mut State, &mut Term, &str) -> Res<()>,
    hide_menu: bool,
) -> Action {
    create_prompt_with_default(prompt, callback, |_| None, hide_menu)
}

pub(crate) fn create_prompt_with_default(
    prompt: &'static str,
    callback: fn(&mut State, &mut Term, &str) -> Res<()>,
    default_fn: fn(&State) -> Option<String>,
    hide_menu: bool,
) -> Action {
    Rc::new(move |state: &mut State, _term: &mut Term| {
        set_prompt(
            state,
            prompt,
            invoke_default,
            Box::new(default_fn),
            callback,
            hide_menu,
        );
        Ok(())
    })
}

fn invoke_default(
    state: &mut State,
    term: &mut Term,
    value: &str,
    context: &fn(&mut State, &mut Term, &str) -> Res<()>,
) -> Res<()> {
    context(state, term, value)
}

type DefaultFn = Box<dyn Fn(&State) -> Option<String>>;

pub(crate) fn set_prompt<T: 'static>(
    state: &mut State,
    prompt: &'static str,
    callback: fn(&mut State, &mut Term, &str, &T) -> Res<()>,
    default_fn: DefaultFn,
    context: T,
    hide_menu: bool,
) {
    let prompt_text = if let Some(default) = default_fn(state) {
        format!("{} (default {}):", prompt, default).into()
    } else {
        format!("{}:", prompt).into()
    };

    if hide_menu {
        state.hide_menu();
    }

    state.prompt.set(PromptData {
        prompt_text,
        update_fn: Rc::new(move |state, term| {
            if state.prompt.state.status().is_done() {
                let input = state.prompt.state.value().to_string();
                state.prompt.reset(term)?;

                let default_value = default_fn(state);
                let value = match (input.as_str(), &default_value) {
                    ("", None) => "",
                    ("", Some(selected)) => selected,
                    (value, _) => value,
                };

                callback(state, term, value, &context)?;

                if hide_menu {
                    state.unhide_menu();
                }
            }
            Ok(())
        }),
    });
}

pub(crate) fn selected_rev(state: &State) -> Option<String> {
    match &state.screen().get_selected_item().target_data {
        Some(TargetData::Branch(branch)) => Some(branch.to_owned()),
        Some(TargetData::Commit(commit)) => Some(commit.to_owned()),
        _ => None,
    }
}
