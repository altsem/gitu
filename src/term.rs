use crate::{error::Error, Res};
use ratatui::{
    backend::{Backend, TestBackend},
    layout::Size,
    prelude::{backend::WindowSize, buffer::Cell, Position, TermwizBackend},
    Terminal,
};
use std::{io, time::Duration};
use termwiz::{
    caps::Capabilities,
    input::InputEvent,
    terminal::Terminal as _,
    terminal::{buffered::BufferedTerminal, SystemTerminal},
};

pub type Term = Terminal<TermBackend>;

pub fn create_backend() -> Res<TermBackend> {
    let buffered_terminal = BufferedTerminal::new(
        SystemTerminal::new(Capabilities::new_from_env().map_err(Error::Termwiz)?)
            .map_err(Error::Termwiz)?,
    )
    .map_err(Error::Termwiz)?;

    Ok(TermBackend::Termwiz(
        TermwizBackend::with_buffered_terminal(buffered_terminal),
    ))
}

pub enum TermBackend {
    Termwiz(TermwizBackend),
    #[allow(dead_code)]
    Test {
        backend: TestBackend,
        events: Vec<InputEvent>,
    },
}

impl Backend for TermBackend {
    fn draw<'a, I>(&mut self, content: I) -> io::Result<()>
    where
        I: Iterator<Item = (u16, u16, &'a Cell)>,
    {
        match self {
            TermBackend::Termwiz(t) => t.draw(content),
            TermBackend::Test { backend, .. } => backend.draw(content),
        }
    }

    fn hide_cursor(&mut self) -> io::Result<()> {
        match self {
            TermBackend::Termwiz(t) => t.hide_cursor(),
            TermBackend::Test { backend, .. } => backend.hide_cursor(),
        }
    }

    fn show_cursor(&mut self) -> io::Result<()> {
        match self {
            TermBackend::Termwiz(t) => t.show_cursor(),
            TermBackend::Test { backend, .. } => backend.show_cursor(),
        }
    }

    fn get_cursor_position(&mut self) -> io::Result<Position> {
        match self {
            TermBackend::Termwiz(t) => t.get_cursor_position(),
            TermBackend::Test { backend, .. } => backend.get_cursor_position(),
        }
    }

    fn set_cursor_position<P: Into<Position>>(&mut self, position: P) -> io::Result<()> {
        match self {
            TermBackend::Termwiz(t) => t.set_cursor_position(position),
            TermBackend::Test { backend, .. } => backend.set_cursor_position(position),
        }
    }

    fn clear(&mut self) -> io::Result<()> {
        match self {
            TermBackend::Termwiz(t) => t.clear(),
            TermBackend::Test { backend, .. } => backend.clear(),
        }
    }

    fn size(&self) -> io::Result<Size> {
        match self {
            TermBackend::Termwiz(t) => t.size(),
            TermBackend::Test { backend, .. } => backend.size(),
        }
    }

    fn window_size(&mut self) -> io::Result<WindowSize> {
        match self {
            TermBackend::Termwiz(t) => t.window_size(),
            TermBackend::Test { backend, .. } => backend.window_size(),
        }
    }

    fn flush(&mut self) -> io::Result<()> {
        match self {
            TermBackend::Termwiz(t) => t.flush(),
            TermBackend::Test { backend, .. } => backend.flush(),
        }
    }
}

impl TermBackend {
    pub fn enter_alternate_screen(&mut self) -> Res<()> {
        match self {
            TermBackend::Termwiz(c) => c
                .buffered_terminal_mut()
                .terminal()
                .enter_alternate_screen()
                .map_err(Error::Termwiz),
            TermBackend::Test { .. } => Ok(()),
        }
    }

    pub fn enable_raw_mode(&mut self) -> Res<()> {
        match self {
            TermBackend::Termwiz(t) => t
                .buffered_terminal_mut()
                .terminal()
                .set_raw_mode()
                .map_err(Error::Termwiz),
            TermBackend::Test { .. } => Ok(()),
        }
    }

    pub fn disable_raw_mode(&mut self) -> Res<()> {
        match self {
            TermBackend::Termwiz(t) => t
                .buffered_terminal_mut()
                .terminal()
                .set_cooked_mode()
                .map_err(Error::Termwiz),
            TermBackend::Test { .. } => Ok(()),
        }
    }

    pub fn poll_input(&mut self, wait: Option<Duration>) -> Res<Option<InputEvent>> {
        match self {
            TermBackend::Termwiz(t) => t
                .buffered_terminal_mut()
                .terminal()
                .poll_input(wait)
                .map_err(Error::Termwiz),
            TermBackend::Test { events, .. } => events.pop().map(Some).ok_or(Error::NoMoreEvents),
        }
    }
}
