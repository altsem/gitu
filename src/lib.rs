pub mod cli;
pub mod config;
mod git;
mod git2_opts;
mod items;
mod keybinds;
mod ops;
mod prompt;
mod screen;
pub mod state;
pub mod term;
#[cfg(test)]
mod tests;
mod ui;

use crossterm::event::{self};
use git2::Repository;
use items::Item;
use itertools::Itertools;
use ops::{Action, Op, SubmenuOp};
use state::State;
use std::{borrow::Cow, error::Error, iter, path::PathBuf, process::Command, rc::Rc};
use term::Term;

const APP_NAME: &str = "gitu";

pub type Res<T> = Result<T, Box<dyn Error>>;

pub(crate) struct CmdMetaBuffer {
    pub(crate) args: Cow<'static, str>,
    pub(crate) out: Option<String>,
}

pub(crate) struct ErrorBuffer(String);

fn command_args(cmd: &Command) -> Cow<'static, str> {
    iter::once(cmd.get_program().to_string_lossy())
        .chain(cmd.get_args().map(|arg| arg.to_string_lossy()))
        .join(" ")
        .into()
}

pub fn run(args: &cli::Args, term: &mut Term) -> Res<()> {
    log::debug!("Finding git dir");
    let dir = PathBuf::from(
        String::from_utf8(
            Command::new("git")
                .args(["rev-parse", "--show-toplevel"])
                .output()?
                .stdout,
        )?
        .trim_end(),
    );

    log::debug!("Opening repo");
    let repo = Repository::open_from_env()?;
    repo.set_workdir(&dir, false)?;

    log::debug!("Initializing config");
    let config = config::init_config()?;

    log::debug!("Creating initial state");
    let mut state = state::State::create(repo, term.size()?, args, config)?;

    log::debug!("Drawing initial frame");
    term.draw(|frame| ui::ui(frame, &mut state))?;

    if args.print {
        return Ok(());
    }

    while !state.quit {
        log::debug!("Awaiting event");
        let event = event::read()?;

        log::debug!("Updating");
        state.update(term, &[event])?;
    }

    Ok(())
}

pub(crate) fn handle_op(state: &mut State, op: Op, term: &mut Term) -> Res<()> {
    let target = state.screen().get_selected_item().target_data.as_ref();
    if let Some(mut action) = op.implementation().get_action(target) {
        Rc::get_mut(&mut action).unwrap()(state, term)?;

        close_submenu(state, op);
    }

    Ok(())
}

fn close_submenu(state: &mut State, op: Op) {
    match op {
        Op::Submenu(_) => (),
        _ => state.pending_submenu_op = SubmenuOp::None,
    }
}
