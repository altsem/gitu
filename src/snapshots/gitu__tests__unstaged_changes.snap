---
source: src/lib.rs
expression: "redact_hashes(terminal, &state.repo)"
---
Buffer {
    area: Rect { x: 0, y: 0, width: 60, height: 20 },
    content: [
        "ðŸ¢’On branch main                                             ",
        " Your branch is up to date with 'origin/main'.              ",
        "                                                            ",
        " Unstaged changes (1)                                       ",
        " testfile                                                   ",
        " @@ -1,2 +1,2 @@                                            ",
        " -testing                                                   ",
        " +test                                                      ",
        "  testtest                                                  ",
        "                                                            ",
        " Recent commits                                             ",
        " _______ main add testfile                                  ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
        "                                                            ",
    ],
    styles: [
        x: 0, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 1, y: 0, fg: Rgb(216, 166, 87), bg: Rgb(80, 73, 69), underline: Reset, modifier: BOLD,
        x: 15, y: 0, fg: Reset, bg: Rgb(80, 73, 69), underline: Reset, modifier: NONE,
        x: 0, y: 1, fg: Reset, bg: Rgb(42, 40, 39), underline: Reset, modifier: NONE,
        x: 0, y: 2, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 3, fg: Rgb(216, 166, 87), bg: Reset, underline: Reset, modifier: BOLD,
        x: 21, y: 3, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 4, fg: Rgb(211, 134, 155), bg: Reset, underline: Reset, modifier: BOLD,
        x: 9, y: 4, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 5, fg: Rgb(125, 174, 163), bg: Reset, underline: Reset, modifier: NONE,
        x: 16, y: 5, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 6, fg: Rgb(234, 105, 98), bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 6, fg: Rgb(234, 105, 98), bg: Reset, underline: Reset, modifier: REVERSED,
        x: 9, y: 6, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 7, fg: Rgb(169, 182, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 2, y: 7, fg: Rgb(169, 182, 101), bg: Reset, underline: Reset, modifier: REVERSED,
        x: 6, y: 7, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 10, fg: Rgb(216, 166, 87), bg: Reset, underline: Reset, modifier: BOLD,
        x: 15, y: 10, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 1, y: 11, fg: Rgb(216, 166, 87), bg: Reset, underline: Reset, modifier: NONE,
        x: 8, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
        x: 9, y: 11, fg: Rgb(169, 182, 101), bg: Reset, underline: Reset, modifier: NONE,
        x: 13, y: 11, fg: Reset, bg: Reset, underline: Reset, modifier: NONE,
    ]
}
